pkg/slo-controller/noderesource/plugins/cpunormalization/plugin.go

```diff
@@ -0,0 +1,293 @@
+}
+
+// Calculate retrieves the CPUNormalizationStrategy from ConfigMap and the CPUBasicInfo from the NRT, get the cpu
+// normalization ratio annotation
+func (p *Plugin) Calculate(_ *configuration.ColocationStrategy, node *corev1.Node, _ *corev1.PodList, _ *framework.ResourceMetrics) ([]framework.ResourceItem, error) {
+	// If any of the necessary inputs like config, CPUBasicInfo is missing, abort to output and skip updating the
+	// node annotation.
+	if !cfgHandler.IsCfgAvailable() {
+		return nil, fmt.Errorf("failed to get config in cpu normalization calculation, err: config is not available")
+	}
+	strategy := cfgHandler.GetStrategyCopy(node)
+	nodeEnabled, err := extension.GetCPUNormalizationEnabled(node)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node label in cpu normalization calculation, err: %w", err)
+	}
+
+	// node label take precedence to the strategy, <nodeEnabled, strategyEnabled>:
+	// - <true, strategyEnabled> -> true
+	// - <false, strategyEnabled> -> false
+	// - <nil, true> -> true
+	// - otherwise -> false
+	if nodeEnabled != nil && !*nodeEnabled || nodeEnabled == nil && (strategy.Enable == nil || strategy.Enable != nil && !*strategy.Enable) {
+		klog.V(6).Infof("cpu normalization is disabled for node %s, reset to ratio %s", node.Name, defaultRatioStr)
+		return []framework.ResourceItem{
+			{
+				Annotations: map[string]string{
+					extension.AnnotationCPUNormalizationRatio: defaultRatioStr,
+				},
+			},
+		}, nil
+	}
+
+	nrt := &topologyv1alpha1.NodeResourceTopology{}
+	err = client.Get(context.TODO(), types.NamespacedName{Name: node.Name}, nrt)
+	if err != nil || nrt == nil {
+		return nil, fmt.Errorf("failed to get NodeResourceTopology in cpu normalization calculation, err: %w", err)
+	}
+	basicInfo, err := extension.GetCPUBasicInfo(nrt.Annotations)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse CPUBasicInfo in cpu normalization calculation, err: %w", err)
+	}
+	if basicInfo == nil {
+		return nil, fmt.Errorf("failed to get CPUBasicInfo in cpu normalization calculation, err: info is missing")
+	}
+
+	ratio, err := getCPUNormalizationRatioFromModel(basicInfo, strategy)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ratio in cpu normalization calculation, err: %s", err)
+	}
+	if err = isCPUNormalizationRatioValid(ratio); err != nil {
+		return nil, fmt.Errorf("failed to validate ratio in cpu normalization calculation, ratio %v, err: %s", ratio, err)
+	}
+
+	ratioStr := strconv.FormatFloat(ratio, 'f', 2, 64)
+	klog.V(6).Infof("calculate cpu normalization ratio %s(%v) for node %s", ratioStr, ratio, node.Name)
+	return []framework.ResourceItem{
+		{
+			Annotations: map[string]string{
+				extension.AnnotationCPUNormalizationRatio: ratioStr,
+			},
+		},
+	}, nil
+}
+
+// MutateNodeResource mutates the NodeResource to amplify the other extended resources like batch-cpu.
+func (p *Plugin) MutateNodeResource(nr *framework.NodeResource, node *corev1.Node) (bool, error) {
```

Comment:
it seems Prepare can be run after batch resource plugin

Reply:
In the current version, the final update `node.status.allocatable` behavior is done by the BatchResource plugin, so the CPUNormalization plugin needs to mutate before the BatchResource.

Reply:
seems the final step "NodeResource->Node" should not be done in plugin, but the controller. leave a todo here

Reply:
done